packages/test/test-end-to-end-tests/src/test/gc/gcTrailingOps.spec.ts

```diff
@@ -17,200 +17,230 @@ import {
-			mainDefaultDataStore._root.set("datastore", newDataStore.entryPoint);
-			mainDefaultDataStore._root.set("blob", blobHandle);
-			mainDefaultDataStore._root.delete("datastore");
-			mainDefaultDataStore._root.delete("blob");
-
-			// Summarize and verify that the datastore and blob are unreferenced
+			// Summarize again to ensure that GC sweep op (if any) is now processed.
 			await provider.ensureSynchronized();
-			const summary1 = await summarizeNow(mainSummarizer);
-			const unreferencedTimestamps1 = await getUnreferencedTimestamps(summary1.summaryTree);
-			const dataStoreTimestamp1 = unreferencedTimestamps1.get(
-				newDataStore.entryPoint.absolutePath.slice(1),
+			const summary3 = await summarizeNow(summarizer);
+
+			// Validate that data store is still in the same state as before and is not deleted.
+			validateDataStoreStateInSummary(
+				summary3.summaryTree,
+				newDataStore.entryPoint.absolutePath,
+				true /* expectGCStateHandle */,
+				transitionToReference,
 			);
-			const blobTimestamp1 = unreferencedTimestamps1.get(blobHandle.absolutePath.slice(1));
-			assert(dataStoreTimestamp1 !== undefined, `Should have unreferenced datastore`);
-			assert(blobTimestamp1 !== undefined, `Should have unreferenced blob`);
-
-			// Create trailing ops where the datastore and blob are referenced
-			mainDefaultDataStore._root.set("datastore", newDataStore.entryPoint);
-			mainDefaultDataStore._root.set("blob", blobHandle);
-			await provider.ensureSynchronized();
-
-			mainContainer.close();
-			mainSummarizer.close();
-
-			// Load a new container/summarizer from the summary and trailing ops
-			const { summarizer } = await createSummarizer(
-				provider,
-				mainContainer,
-				{ runtimeOptions: { gcOptions }, loaderProps: { configProvider } },
-				summary1.summaryVersion,
-			);
-
-			// Ensure trailing ops are processed, summarize, and verify that the datastore and blob are referenced
-			await provider.ensureSynchronized();
-			const summary2 = await summarizeNow(summarizer);
-			const unreferencedTimestamps2 = await getUnreferencedTimestamps(summary2.summaryTree);
-			const dataStoreId = newDataStore.entryPoint.absolutePath.slice(1);
-			const blobId = blobHandle.absolutePath.slice(1);
-			assert(unreferencedTimestamps2.has(dataStoreId), `GC should detect the datastore`);
-			assert(unreferencedTimestamps2.has(blobId), `GC should detect the blob`);
-			const dataStoreTimestamp2 = unreferencedTimestamps2.get(dataStoreId);
-			const blobTimestamp2 = unreferencedTimestamps2.get(blobId);
-			assert(dataStoreTimestamp2 === undefined, `Should have a referenced datastore`);
-			assert(blobTimestamp2 === undefined, `Should have a referenced blob`);
 		});
 	};
 
-	tests();
-	tests(true /** tombstoneEnabled */);
+	tests(true /* transitionToReference */, true /** beforeSweepTimeout */);
+	tests(true /* transitionToReference */, false /** beforeSweepTimeout */);
+	tests(false /** transitionToReference */, true /* beforeSweepTimeout */);
```

Comment:
In this case something transitions to unreferenced and then the sweep timeout elapses and then GC runs.  Naively it seems like it should be deleted!

Reply:
Like here's the timeline where 100 is sweep timeout:
* T0 - Summary1 - it's referenced.
* T1 - op unrefrences it.  This will be the unreferenced time since no other ops will go in Summary2
* T101 - Summary2 - it's SweepReady (right?)
* T102 - Summary3 - Delete op is sent

So if you did _another_ summary wouldn't it be gone?

Reply:
yeah, that's a side-effect of the fact that the reference state of objects is updated when GC runs and not immediately.

Reply:
> Like here's the timeline where 100 is sweep timeout:
> 
> * T0 - Summary1 - it's referenced.
> * T1 - op unrefrences it.  This will be the unreferenced time since no other ops will go in Summary2
> * T101 - Summary2 - it's SweepReady (right?)
> * T102 - Summary3 - Delete op is sent
> 
> So if you did _another_ summary wouldn't it be gone?

At T101, Summary2 will not use T1 but the timestamp of the summarizer's join op. But more importantly, the data store won't even be unreferenced until Summary2. So, it doesn't matter how much time has passed before that.

Reply:
Ah, right, I forgot about the join op.  I was thinking Summary2's latest op timestamp would be way back at T1.